### PR TITLE
Perform connection

### PR DESCRIPTION
**Pull request by jpeter96**
### **User description**
Implemented: "Testing The Client" (on MS01 code-base) pl review 
Still to be done: Test performConnection with LRZ-Server

___

### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Refactored the `performConnection` function in `performConnection.c` to use consistent return values (`0` and `-1`) for success and failure.
- Added a comprehensive test suite for the `args_parser` module, covering valid, invalid, and edge-case inputs.
- Implemented unit tests for the `createConnection` function using `cmocka`, including scenarios like successful connection, hostname resolution failure, and memory cleanup.
- Updated the project structure in `README.md` to reflect new test directories and files.
- Added detailed documentation for the `args_parser` module and its test suite, including usage, compilation, and test cases.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>performConnection.c</strong><dd><code>Refactored return values in `performConnection` function.</code></dd></summary>
<hr>

modules/tcp_performConnection/performConnection.c

<li>Adjusted return values in <code>performConnection</code> function to use <code>0</code> and <code>-1</code> <br>instead of <code>EXIT_SUCCESS</code> and <code>EXIT_FAILURE</code>.<br> <li> Improved consistency in error handling and return codes.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/23/files#diff-984e74d1b03c4aafe88a26e4b34a74cf255bbc04741004fc1751bf82cc522357">+6/-7</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_args_parser.c</strong><dd><code>Added test suite for `args_parser` module.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_args_parser/test_args_parser.c

<li>Added a comprehensive test suite for the <code>args_parser</code> module.<br> <li> Included multiple test cases for valid, invalid, and edge-case inputs.<br> <li> Validates functionality of <code>parse_args</code> and configuration extraction.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/23/files#diff-d2bf3ed88b7b2d28c0a09ef595ecbc695361085bc94d5451472baf20d89c9d97">+105/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_create_tcp_connection.c</strong><dd><code>Added unit tests for `createConnection` function.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_tcp_connection/test_create_tcp_connection.c

<li>Implemented unit tests for <code>createConnection</code> function using <code>cmocka</code>.<br> <li> Mocked functions like <code>socket</code>, <code>connect</code>, and <code>performConnection</code>.<br> <li> Covered scenarios such as successful connection, hostname resolution <br>failure, and memory cleanup.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/23/files#diff-cf7d4977d1a6c6c560dad0a376b0df1a05c3dfa8c9f330dcecb5125b4c547922">+133/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Updated project structure documentation.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

README.md

<li>Updated project structure to include new test directories and files.<br> <li> Documented additional test modules and their locations.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/23/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+10/-5</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Documented `args_parser` module and test suite.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_args_parser/README.md

<li>Added documentation for the <code>args_parser</code> module and its test suite.<br> <li> Included details on usage, compilation, and test cases.<br> <li> Highlighted validation rules and edge cases for command-line <br>arguments.<br>

</details>

  </td>
  <td><a href="https://github.com/noluyorAbi/C-Project/pull/23/files#diff-81d545bb7c980df4d913edd60525e968b8066dd9d97593f38aef1572b36a6405">+61/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by noluyorAbi** [comment]
Preparing review...

**Comment by noluyorAbi** [comment]
Preparing review...

**Comment by noluyorAbi** [comment]
Preparing review...

**Comment by noluyorAbi** [comment]
## PR Reviewer Guide 🔍

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 3 🔵🔵🔵⚪⚪</td></tr>
<tr><td>🧪&nbsp;<strong>PR contains tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Key issues to review</strong><br><br>

<a href='https://github.com/noluyorAbi/C-Project/pull/23/files#diff-984e74d1b03c4aafe88a26e4b34a74cf255bbc04741004fc1751bf82cc522357R72-R73'><strong>Consistency Issue</strong></a><br>The return values for error handling in `performConnection` function have been changed from `EXIT_FAILURE` to `-1`. It is recommended to use standard exit codes for consistency and clarity.

<a href='https://github.com/noluyorAbi/C-Project/pull/23/files#diff-984e74d1b03c4aafe88a26e4b34a74cf255bbc04741004fc1751bf82cc522357R109-R110'><strong>Consistency Issue</strong></a><br>The return values for error handling in `performConnection` function have been changed from `EXIT_FAILURE` to `-1`. It is recommended to use standard exit codes for consistency and clarity.

<a href='https://github.com/noluyorAbi/C-Project/pull/23/files#diff-984e74d1b03c4aafe88a26e4b34a74cf255bbc04741004fc1751bf82cc522357R145-R146'><strong>Consistency Issue</strong></a><br>The return values for error handling in `performConnection` function have been changed from `EXIT_FAILURE` to `-1`. It is recommended to use standard exit codes for consistency and clarity.

</td></tr>
</table>

File: modules/tcp_performConnection/performConnection.c [hunk 1]
```diff
@@ -70,8 +69,8 @@ int performConnection(int sockfd) {
 
   // 2. Send client version
   const char *client_version = "VERSION 2.42\n";
-  if (sendMessage(sockfd, client_version) != EXIT_SUCCESS) {
-    return EXIT_FAILURE;
+  if (sendMessage(sockfd, client_version) != 0) {
+    return -1;
```
Comment (noluyorAbi) [inline modules/tcp_performConnection/performConnection.c:73]:
```suggestion
    return EXIT_FAILURE;
```
We use EXIT_FAILURE & EXIT_SUCCESS as  standard

File: modules/tcp_performConnection/performConnection.c [hunk 2]
```diff
@@ -107,8 +106,8 @@ int performConnection(int sockfd) {
 
   // 7. Send PLAYER command (no additional values)
   const char *player_command = "PLAYER\n";
-  if (sendMessage(sockfd, player_command) != EXIT_SUCCESS) {
-    return EXIT_FAILURE;
+  if (sendMessage(sockfd, player_command) != 0) {
+    return -1;
```
Comment (noluyorAbi) [inline modules/tcp_performConnection/performConnection.c:110]: here too

File: modules/tcp_performConnection/performConnection.c [hunk 3]
```diff
@@ -143,8 +142,8 @@ int performConnection(int sockfd) {
 
   // 10. Receive details of other players
   while (1) {
-    if (receiveMessage(sockfd, buffer, BUFFER_SIZE) != EXIT_SUCCESS) {
-      return EXIT_FAILURE;
+    if (receiveMessage(sockfd, buffer, BUFFER_SIZE) != 0) {
```
Comment (noluyorAbi) [inline modules/tcp_performConnection/performConnection.c:145]: see above

**Comment by noluyorAbi** [comment]
I implemented some comments and adjustments, but I couldn’t get `test_create_tcp_connection.c` to work even after several attempts. Did it work for you?

Maybe we should add Makefiles to the folders to ensure the tests run correctly, even for the tutors reviewing our code.

File: tests/test_args_parser/test_args_parser.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
+#include "../../modules/args_parser/args_parser.h"
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+// Hilfsfunktion, um Kommandozeilenargumente zu simulieren
```
[the remark this answers is not in the document]
Comment (noluyorAbi) [reply tests/test_args_parser/test_args_parser.c:7]: Readmes can stay in german

**Review by noluyorAbi — CHANGES_REQUESTED** [review]

**Comment by cininja** [comment]
This pull request will be closed, as numerous changes and tests are already implemented on a different branch, making most of the adjustments in this request likely redundant.
